webhook/aws/sns.go

```diff
@@ -15,3 +28,120 @@ type SNSConfig struct {
 	TopicArn string `json:"topicArn"`
 	UrlPath  string `json:"urlPath"` //uri path to register mux
 }
+
+type SNSServer struct {
+	Config          *AWSConfig
+	SubscriptionArn string
+	SVC             *sns.SNS
+	SelfUrl         *url.URL
+	Logger			logging.Logger
+	// mutex used to protect SubscriptionArn read & write
+	sync.RWMutex
+}
+
+// NewSNSServer initializes the SNSServer
+// selfURL represents the webhook server URL &url.URL{Scheme:secure,Host:fqdn+port,Path:urlPath}
+// handler is the webhook handler to update webhooks @monitor 
+// SNS POST Notification handler will directly update webhooks list
+
+func NewSNSServer(cfg *AWSConfig, logger logging.Logger, rtr *mux.Router, 
+	selfUrl *url.URL, handler http.Handler) (ss *SNSServer, err error) {
+	
+	if cfg == nil {
+		return nil, fmt.Errorf("Invalid AWS Config")
+	}
+	
+	if rtr == nil {
+		//creating new mux router
+		rtr = mux.NewRouter()
+	}
+	
+	cred := credentials.NewStaticCredentials(cfg.AccessKey, cfg.SecretKey, "")
+	
+	sess, aws_err := session.NewSession(&aws.Config{
+                Region:      aws.String(cfg.Sns.Region),
+                Credentials: cred,
+        })
+	if aws_err != nil {
+		return nil, aws_err
+	}
+
+	svc := sns.New(sess)
+
+	// Set webhook url path to SNS UrlPath
+	selfUrl.Path = cfg.Sns.UrlPath
+
+	// Initialize the server
+	ss = &SNSServer{
+		Config:   cfg,
+		SVC:      svc,
+		SelfUrl:  selfUrl,
+	}
+	
+	// set up logger
+	ss.Logger = logger
+	
+	ss.logger().Debug("SNS self url endpoint: [%s], protocol [%s]", ss.SelfUrl.String(), ss.SelfUrl.Scheme)
+
+	// Set various SNS POST routes
+	ss.SetSNSRoutes(cfg.Sns.UrlPath, rtr, handler)
+	
+	return ss, nil
+}
+
+func (ss *SNSServer) logger() logging.Logger {
+	if ss != nil && ss.Logger != nil {
+		return ss.Logger
+	}
+	return logging.DefaultLogger()
+}
+
+// Prepare the SNSServer to receive Notifications 
+// This better be called after the endpoint http server is started 
+// and ready to receive AWS SNS POST messages
+// subscribe to the SNS topic, wait for snsReady
+// validate the confirmation SubscriptionArn
+func (ss *SNSServer) PrepareAndStart() {
+	ss.Subscribe()
+}
+
+// Returns true if the SNS is ready to accept notifications
+// Synchronized using read lock
+func (ss *SNSServer) IsReady() bool {
+	var ready bool
+	ss.RLock()
+	if !strings.EqualFold("pending confirmation", "") && !strings.EqualFold("pending confirmation", ss.SubscriptionArn) {	
```

Comment:
Oh yes you are right, somehow it never occurred to that I am comparing 2 const strings